File: loo.ts
import {existsSync, readFileSync, statSync} from 'fs';
import {argv, stdout} from 'process';

type Characters = '+' | '-' | '#' | ';';

function execute(data: string) {
  const currentCharacter = (pos: number, source: string): string | null => {
    if (source.length == pos) {
      return null;
    } else {
      return source[pos];
    }
  };

  let position: number = 0;
  let ascii: number = 0;
  let character: string | null = currentCharacter(position, data);
  while (character != null) {
    if (character == '+') {
      ascii += 1;
    } else if (character == '-') {
      ascii -= 1;
    } else if (character == '#') {
      stdout.write(String.fromCharCode(ascii));
    } else if (character == ';') {
      return 0;
    }

    position += 1;
    character = currentCharacter(position, data);
  }
}

function content(args: Array<string>): string | null {
  if (args.length == 0) {
    return null;
  }

  const filename: string = arguments[0][0];
  try {
    const exists = existsSync(filename);
    if (!statSync(filename).isFile()) {
      console.error(`${filename} not a file`);
    }
  } catch (exception: any) {
    console.error(exception.message);
  }
  const data = readFileSync(filename).toString();
  return data;
}

const data: string | null = content(argv.slice(2));
if (data) {
  execute(data);
}

File: loo.dart
import 'dart:io';

class LooInterpreter {
  int position = 0;
  int ascii = 0;

  late String data;

  LooInterpreter(String data) {
    this.data = data;
  }

  void execute() {
    String character = this.currentCharacter();
    while(character != ";"){
      if(character == "+"){
        this.ascii += 1;
      } else if(character == "-"){
        this.ascii -= 1;
      } else if(character == "#") {
        stdout.write(String.fromCharCode(this.ascii));
      }

      this.position += 1;
      character = this.currentCharacter();
    }
  }

  String currentCharacter() {
    if(this.data.length == this.position) {
      return ";";
    } else {
      return this.data[this.position];
    }
  }
}

String createSource(List<String> arguments){
  if(arguments.length == 0){
    return "";
  }
  String filename = arguments[0];
  return filename;
}

void main(List<String> arguments) {
  String filename = createSource(arguments);
  if(filename.length == 0){
    print("No filename specified");
  } else {
    File(filename).readAsString().then((String content) {
      LooInterpreter interpreter = new LooInterpreter(content);
      interpreter.execute();
    });
  }
}
